localstack/services/awslambda/invocation/runtime_executor.py

```diff
@@ -0,0 +1,233 @@
+from localstack.utils.docker_utils import DOCKER_CLIENT as CONTAINER_CLIENT
+from localstack.utils.net import get_free_tcp_port
+
+LOG = logging.getLogger(__name__)
+
+RUNTIME_REGEX = r"(?P<runtime>[a-z]+)(?P<version>\d+(\.\d+)?(\.al2)?)(?:.*)"
+
+# IMAGE_PREFIX = "gallery.ecr.aws/lambda/"
+IMAGE_PREFIX = "amazon/aws-lambda-"
+
+RAPID_ENTRYPOINT = "/var/rapid/init"
+
+InitializationType = Literal["on-demand", "provisioned-concurrency"]
+
+LAMBDA_DOCKERFILE = """FROM {base_img}
+COPY aws-lambda-rie {rapid_entrypoint}
+COPY code/ /var/task
+"""
+
+
+# TODO provided runtimes
+# TODO a tad hacky, might cause problems in the future.. just use mapping?
+def get_runtime_split(runtime: str) -> Tuple[str, str]:
+    match = re.match(RUNTIME_REGEX, runtime)
+    if match:
+        runtime, version = match.group("runtime"), match.group("version")
+        # sad exception for .net
+        if runtime == "dotnetcore":
+            runtime = "dotnet"
+            version = f"core{version}"
+        return runtime, version
+    raise Exception("Cannot process runtime '%s'" % runtime)
+
+
+def get_path_for_function(function_version: FunctionVersion) -> Path:
+    return Path(
+        f"{config.dirs.tmp}/lambda/{function_version.id.qualified_arn().replace(':', '_').replace('$', '_')}/"
+    )
+
+
+def get_code_path_for_function(function_version: FunctionVersion) -> Path:
+    return get_path_for_function(function_version) / "code"
+
+
+def get_image_name_for_function(function_version: FunctionVersion) -> str:
+    return f"localstack/lambda-{function_version.id.qualified_arn().replace(':', '_').replace('$', '_').lower()}"
+
+
+def get_image_for_runtime(runtime: str) -> str:
+    runtime, version = get_runtime_split(runtime)
+    return f"{IMAGE_PREFIX}{runtime}:{version}"
+
+
+def get_runtime_client_path() -> Path:
+    return Path(LAMBDA_RUNTIME_INIT_PATH)
+
+
+def prepare_image(target_path: Path, function_version: FunctionVersion) -> None:
+    if not function_version.config.runtime:
+        LOG.error("Images without runtime are currently not supported")
+        raise Exception("Custom images are currently not supported")
+    src_init = get_runtime_client_path()
+    # copy init file
+    target_init = target_path / "aws-lambda-rie"
+    shutil.copy(src_init, target_init)
+    target_init.chmod(0o755)
+    # copy code
+    # create dockerfile
+    docker_file_path = target_path / "Dockerfile"
+    docker_file = LAMBDA_DOCKERFILE.format(
+        base_img=get_image_for_runtime(function_version.config.runtime),
+        rapid_entrypoint=RAPID_ENTRYPOINT,
+    )
+    with docker_file_path.open(mode="w") as f:
+        f.write(docker_file)
+    try:
+        CONTAINER_CLIENT.build_image(
+            dockerfile_path=str(docker_file_path),
+            image_name=get_image_name_for_function(function_version),
+        )
+    except Exception as e:
+        LOG.error("Exception: %s", e)
```

Comment:
a good pattern is

```python
if LOG.isEnabledFor(logging.DEBUG):
    LOG.exception("error while ...")
else:
    LOG.error("error while ...: %s", e)
```

`LOG.exception` prints the stacktrace from the current frame, no need to pass `e`